crates/accelerate/src/basis/basis_translator/basis_search.rs

```diff
@@ -0,0 +1,228 @@
+            match event {
+                DijkstraEvent::Discover(n, score) => {
+                    let gate_key = &equiv_lib.graph()[n].key;
+                    let gate = (gate_key.name.to_string(), gate_key.num_qubits);
+                    source_basis_remain.remove(gate_key);
+                    let mut borrowed_cost_map = opt_cost_map.borrow_mut();
+                    if let Some(entry) = borrowed_cost_map.get_mut(&gate) {
+                        *entry = score;
+                    } else {
+                        borrowed_cost_map.insert(gate.clone(), score);
+                    }
+                    if let Some(rule) = predecessors.borrow().get(&gate) {
+                        // TODO: Logger
+                        basis_transforms.push((
+                            (gate_key.name.to_string(), gate_key.num_qubits),
+                            (rule.params.clone(), rule.circuit.clone()),
+                        ));
+                    }
+
+                    if source_basis_remain.is_empty() {
+                        basis_transforms.reverse();
+                        return Control::Break(());
+                    }
+                }
+                DijkstraEvent::EdgeRelaxed(_, target, Some(edata)) => {
+                    let gate = &equiv_lib.graph()[target].key;
+                    predecessors
+                        .borrow_mut()
+                        .entry((gate.name.to_string(), gate.num_qubits))
+                        .and_modify(|value| *value = edata.rule.clone())
+                        .or_insert(edata.rule.clone());
+                }
+                DijkstraEvent::ExamineEdge(_, target, Some(edata)) => {
+                    num_gates_remaining_for_rule
+                        .entry(edata.index)
+                        .and_modify(|val| *val -= 1)
+                        .or_insert(0);
+                    let target = &equiv_lib.graph()[target].key;
+
+                    // If there are gates in this `rule` that we have not yet generated, we can't apply
+                    // this `rule`. if `target` is already in basis, it's not beneficial to use this rule.
+                    if num_gates_remaining_for_rule[&edata.index] > 0
+                        || target_basis_keys.contains(target)
+                    {
+                        return Control::Prune;
+                    }
+                }
+                _ => {}
+            };
+            Control::Continue
+        },
+    ) {
+        Ok(Control::Break(_)) => Some(basis_transforms),
+        _ => None,
+    };
+
+    // TODO: Values will have to be cloned in order for the dummy node to be removed.
+    equiv_lib.mut_graph().remove_node(dummy);
+    basis_transforms
+}
+
+fn initialize_num_gates_remain_for_rule(
+    graph: &StableDiGraph<NodeData, Option<EdgeData>>,
+    source: &mut HashMap<usize, usize>,
+) {
+    let mut save_index = usize::MAX;
+    for edge_data in graph.edge_weights().flatten() {
```

Comment:
Maybe leave a comment here that `flatten()` is for iterating over defined edges and skipping any set with `None`. It took me a second to figure out what this was doing, it's a clever workaround though (`Option<T>` implementing `Iterator` is always confusing to me at first).